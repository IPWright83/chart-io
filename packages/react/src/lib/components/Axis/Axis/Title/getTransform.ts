import type { IMargin, IPosition } from "@chart-it/types";

/**
 * Obtain the transform for the Axis title
 * @param  position     The position of the axis [left, right, top, bottom]
 * @param  width        The width for the chart
 * @param  height       The height for the chart
 * @param  margin       The margin object for the chart
 * @return              The transform
 */
export function getTransform(position: IPosition, width: number, height: number, margin: IMargin): string {
    if (width === 0 || height === 0) {
        return "translate(0, 0)";
    }

    const actualWidth = width - margin.left - margin.right;
    const actualHeight = height - margin.top - margin.bottom;
    const halfWidth = actualWidth / 2;
    const halfHeight = actualHeight / 2;

    switch (position) {
        case "left":
            return `translate(${margin.left / 2 - 25}, ${halfHeight + margin.top / 2})rotate(270)`;
        case "right":
            return `translate(${width - margin.right / 2 + 25}, ${halfHeight + margin.top / 2})rotate(270)`;
        case "top":
            return `translate(${halfWidth + margin.left / 2}, ${margin.top - 25})`;
        case "bottom":
            return `translate(${halfWidth + margin.left / 2}, ${height - margin.bottom + 35})`;
        default:
            throw new Error(`Invalid position: ${position}`);
    }
}
