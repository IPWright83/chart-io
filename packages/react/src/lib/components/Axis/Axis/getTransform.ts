import type { IMargin, IPosition } from "@chart-io/types";
import { logAndThrowError } from "@chart-io/core";

/**
 * Obtain the transform for the Axis group
 * @param  position     The position of the axis [left, right, top, bottom]
 * @param  plotWidth    The width for the plot
 * @param  plotHeight   The height for the plot
 * @param  margin       The margin object for the chart
 * @return              The transform
 */
export function getTransform(position: IPosition, plotWidth: number, plotHeight: number, margin: IMargin): string {
    if (plotWidth === 0 || plotHeight === 0) {
        return "translate(0, 0)";
    }

    switch (position) {
        case "left":
            return `translate(${margin.left}, 0)`;
        case "right":
            return `translate(${plotWidth + margin.left}, 0)`;
        case "top":
            return `translate(0, ${margin.top})`;
        case "bottom":
            return `translate(0, ${plotHeight + margin.top})`;
        default:
            logAndThrowError("E002", `Invalid position (${position}) was provided to the Axis`);
    }
}
