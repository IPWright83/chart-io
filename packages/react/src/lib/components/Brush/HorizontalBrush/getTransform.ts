import type { IMargin } from "@chart-it/types";

/**
 * Gets the transform for the Brush position
 * @param  height   The height of the chart
 * @param  margin   The margin of the chart
 * @return          The transform for the brush
 */
export function getTransform(height: number, margin: IMargin): string {
    return `translate(0, ${height})`;
}
