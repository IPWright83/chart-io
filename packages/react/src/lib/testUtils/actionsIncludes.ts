/**
 * Assert that all the expected dispatch calls to the Redux store were seen
 * @param {string[]} dispatchCalls The actual calls
 * @param {string[]} expectedCalls The expected calls
 */
export function actionsIncludes(dispatchCalls: string[], expectedCalls: string[]) {
    expectedCalls.forEach((expected) => {
        expect(dispatchCalls).toContain(expected);
        dispatchCalls.splice(dispatchCalls.indexOf(expected), 1);
    });
}
