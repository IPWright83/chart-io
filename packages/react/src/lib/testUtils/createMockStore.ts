import { defaultChartState, defaultEventState, IStore } from "../store";

/**
 * Creates a mock Redux store for testing with
 * @param  state     The initial state of the store
 * @return           A mock store
 */
export function createMockStore(state: { chart?: any; event?: any }): IStore {
    return {
        getState: () => ({
            chart: {
                ...{ ...defaultChartState, id: "test" },
                ...(state.chart ?? {}),
            },
            event: {
                ...defaultEventState,
                ...(state.event ?? {}),
            },
        }),
        dispatch: jest.fn(),
        subscribe: jest.fn(),
    } as unknown as IStore;
}

/**
 * Creates a mock Redux store for Storybook
 * @param  state     The initial state of the store
 * @return           A mock store
 */
export function createMockStorybookStore(state: { chart?: any; event?: any }) {
    return {
        getState: () => ({
            chart: {
                ...defaultChartState,
                ...(state.chart ?? {}),
            },
            event: {
                ...defaultEventState,
                ...(state.event ?? {}),
            },
        }),
        //eslint-disable-next-line @typescript-eslint/no-empty-function
        dispatch: () => {},
        //eslint-disable-next-line @typescript-eslint/no-empty-function
        subscribe: () => {},
    } as unknown as IStore;
}
