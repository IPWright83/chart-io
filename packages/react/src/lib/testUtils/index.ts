export * from "./actionsIncludes";
export * from "./FakeMouseEvent";
export * from "./createMockStore";
export * from "./getBuffer";
export * from "./wait";
export * from "./renderChart";
export * from "./testEvents";
