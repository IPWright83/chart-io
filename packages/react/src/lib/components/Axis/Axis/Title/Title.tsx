import type { IPosition } from "@d3-chart/types";
import React from "react";
import { useSelector } from "react-redux";

import { getTransform } from "./getTransform";

import { chartSelectors, IState } from "../../../../store";

export interface ITitleProps {
    /**
     * The position of the title [top, bottom, left, right]
     */
    position: IPosition;
    /**
     * A title for the Axis
     */
    title?: string;
}

export function Title({ position, title }: ITitleProps) {
    const width = useSelector((s: IState) => chartSelectors.dimensions.width(s));
    const height = useSelector((s: IState) => chartSelectors.dimensions.height(s));
    const margin = useSelector((s: IState) => chartSelectors.dimensions.margin(s));

    const transform = getTransform(position, width, height, margin);

    if (!title) {
        return null;
    }

    const style = {
        textAnchor: "middle" as const,
        fontSize: 14,
        userSelect: "none" as const,
    };

    return (
        <text className={`d3-chart axis-title axis-title-${position}`} transform={transform} style={style}>
            {title}
        </text>
    );
}
