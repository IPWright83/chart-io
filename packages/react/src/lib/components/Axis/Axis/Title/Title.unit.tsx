import { Provider } from "react-redux";
import React from "react";
import { render } from "@testing-library/react";

import { createMockStore } from "../../../../testUtils";

import { getTransform } from "./getTransform";
import { Title } from "./Title";

describe("Title", () => {
    const width = 1000;
    const height = 500;
    const margin = { left: 50, right: 60, top: 30, bottom: 40 };

    const store = createMockStore({
        chart: {
            dimensions: {
                width,
                height,
                margin,
            },
        },
    });

    describe("getTransform", () => {
        describe("should return correct transform", () => {
            it("for left axis", () => {
                expect(getTransform("left", width, height, margin)).toBe("translate(0, 230)rotate(270)");
            });
            it("for right axis", () => {
                expect(getTransform("right", width, height, margin)).toBe("translate(995, 230)rotate(270)");
            });
        });

        describe("should return correct transform", () => {
            it("for top axis", () => {
                expect(getTransform("top", width, height, margin)).toBe("translate(470, 5)");
            });
            it("for bottom axis", () => {
                expect(getTransform("bottom", width, height, margin)).toBe("translate(470, 495)");
            });
        });

        describe("should return no translation", () => {
            it("for a 0 width", () => {
                expect(getTransform("top", 0, height, margin)).toBe("translate(0, 0)");
            });
            it("for a 0 height", () => {
                expect(getTransform("top", width, 0, margin)).toBe("translate(0, 0)");
            });
        });

        it("throws an error with an invalid position", () => {
            // @ts-expect-error Testing the case of an invalid value
            expect(() => getTransform("invalid", width, height, margin)).toThrow();
        });
    });

    describe("component", () => {
        describe("with title set", () => {
            it("renders horizontal title", async () => {
                const { asFragment } = render(
                    <Provider store={store}>
                        <svg>
                            <Title position="bottom" title="horizontal" />
                        </svg>
                    </Provider>
                );

                expect(asFragment()).toMatchSnapshot();
            });

            it("renders vertical title", async () => {
                const { asFragment } = render(
                    <Provider store={store}>
                        <svg>
                            <Title position="left" title="vertical" />
                        </svg>
                    </Provider>
                );

                expect(asFragment()).toMatchSnapshot();
            });
        });

        describe("with no explicit title", () => {
            it("renders nothing", async () => {
                const { asFragment } = render(
                    <Provider store={store}>
                        <svg>
                            <Title position="bottom" />
                        </svg>
                    </Provider>
                );

                expect(asFragment()).toMatchSnapshot();
            });
        });
    });
});
