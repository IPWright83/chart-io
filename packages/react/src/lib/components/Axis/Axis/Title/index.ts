import { ITitleProps, Title } from "./Title";

import { getTransform } from "./getTransform";

// @ts-expect-error: extending the interface on purpose
Title.getTransform = getTransform;

export { Title };
export type { ITitleProps };
