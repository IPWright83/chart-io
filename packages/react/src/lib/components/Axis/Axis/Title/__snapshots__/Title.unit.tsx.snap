// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title component with no explicit title renders nothing 1`] = `
<DocumentFragment>
  <svg />
</DocumentFragment>
`;

exports[`Title component with title set renders horizontal title 1`] = `
<DocumentFragment>
  <svg>
    <text
      class="chart-io axis-title axis-title-bottom"
      style="text-anchor: middle; font-size: 14px; user-select: none;"
      transform="translate(470, 495)"
    >
      horizontal
    </text>
  </svg>
</DocumentFragment>
`;

exports[`Title component with title set renders vertical title 1`] = `
<DocumentFragment>
  <svg>
    <text
      class="chart-io axis-title axis-title-left"
      style="text-anchor: middle; font-size: 14px; user-select: none;"
      transform="translate(0, 230)rotate(270)"
    >
      vertical
    </text>
  </svg>
</DocumentFragment>
`;
